File: lib/JFFUtils/NSString/NSString+Search.m
#import "NSString+Search.h"

@implementation NSString (Search)

- (NSUInteger)numberofOccurencesWithRangeSearcher:(NSRange(^)(NSRange))rangeSearcher
                                             step:(NSUInteger)step
{
    NSUInteger result = 0;
    
    NSRange searchRange = {0, [self length]};
    NSRange range = rangeSearcher(searchRange);
    
    while (range.location != NSNotFound)
    {
        ++result;
        
        searchRange.location = range.location + step;
        searchRange.length   = [ self length ] - searchRange.location;
        if (searchRange.location >= [self length])
            break;
        
        range = rangeSearcher(searchRange);
    }
    
    return result;
}

- (NSUInteger)numberOfCharacterFromString:(NSString *)string
{
    NSCharacterSet *set = [NSCharacterSet characterSetWithCharactersInString: string];
    
    NSRange (^rangeSearcher)(NSRange) = ^NSRange(NSRange rangeToSearch)
    {
        return [self rangeOfCharacterFromSet: set
                                     options: NSLiteralSearch
                                       range: rangeToSearch];
    };
    
    return [self numberofOccurencesWithRangeSearcher: rangeSearcher
                                                step: 1];
}

- (NSUInteger)numberOfStringsFromString:(NSString *)string
{
    NSRange (^rangeSearcher)(NSRange) = ^NSRange(NSRange rangeToSearch)
    {
        return [self rangeOfString: string
                           options: NSLiteralSearch
                             range: rangeToSearch];
    };
    
    return [self numberofOccurencesWithRangeSearcher: rangeSearcher
                                                step: [string length]];
}

- (BOOL)containsString:(NSString *)string
{
    NSRange range = [self rangeOfString: string
                                options: NSLiteralSearch
                                  range: (NSRange){0, [self length]}];
    
    return range.location != NSNotFound;
}

- (BOOL)caseInsensitiveContainsString:(NSString *)string
{
    NSRange range = [ self rangeOfString: string
                                 options: NSCaseInsensitiveSearch
                                   range: (NSRange){0, [self length]}];
    
    return range.location != NSNotFound;
}

@end

File: iAsync.Swift/lib/JUtils/JUtils/String/NSString+Search.swift
//
//  NSString+Search.swift
//  JUtils
//
//  Created by Vladimir Gorbenko on 06.06.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

public extension NSString {
    
    private func numberofOccurencesWithRangeSearcher(
        @noescape rangeSearcher: (NSRange) -> NSRange, step: Int) -> Int {
        
        var result = 0
        
        var searchRange = NSMakeRange(0, length)
        var range = rangeSearcher(searchRange)
        
        while range.location != Foundation.NSNotFound {
            
            ++result
            
            searchRange.location = range.location + step
            searchRange.length   = length - searchRange.location
            if searchRange.location >= length {
                break
            }
            
            range = rangeSearcher(searchRange)
        }
        
        return result
    }
    
    func numberOfCharacterFromString(string: String) -> Int {
        
        let set = NSCharacterSet(charactersInString: string)
        
        let rangeSearcher = { (rangeToSearch: NSRange) -> NSRange in
            return self.rangeOfCharacterFromSet(set, options: .LiteralSearch, range: rangeToSearch)
        }
        
        return numberofOccurencesWithRangeSearcher(rangeSearcher, step: 1)
    }
    
    func numberOfStringsFromString(string: String) -> Int {
        
        let rangeSearcher = { (rangeToSearch: NSRange) -> NSRange in
            return self.rangeOfString(string, options: .LiteralSearch, range: rangeToSearch)
        }
        
        let nsStringLength = (string as NSString).length
        
        return numberofOccurencesWithRangeSearcher(rangeSearcher, step: nsStringLength)
    }
    
    func caseInsensitiveContainsString(string: String) -> Bool {
        
        let range = self.rangeOfString(string, options: .CaseInsensitiveSearch, range: NSMakeRange(0, length))
        return range.location != Foundation.NSNotFound
    }
}
